### ISSUE TITLE
Speed up the validity tests

### ISSUE THREAD

**Issue by sjakobi** [commercialhaskell/path#70]
They take nearly two minutes to run on my (admittedly slow) machine.

Maybe they could be parallelized and/or the generation could be sped up.

**Comment by NorfairKing** [comment]
Aha! Looks like we should write our own `genValid` instances.
How do you propose generating only valid paths such that still every valid path can be generated with that generator?

**Comment by sjakobi** [comment]
> How do you propose generating only valid paths such that still every valid path can be generated with that generator?

Good question, I don't know yet.

---

BTW the issue seems quite exacerbated on Windows, see https://github.com/commercialhaskell/path/pull/72#issuecomment-284224611. Might be easier to fix there though.

**Comment by NorfairKing** [comment]
I think we want to decide on #53 first though.

**Comment by sjakobi** [comment]
For more immediate benefits it's probably worth looking into http://hspec.github.io/parallel-spec-execution.html.

**Comment by sjakobi** [comment]
To speed up the generation it might also be possible to use `System.FilePath.makeValid`.

**Comment by NorfairKing** [comment]
> To speed up the generation it might also be possible to use System.FilePath.makeValid.

Perhaps, but we have to make sure that all valid paths can still be generated.

In the docs for `isValid`:

```
         isValid "" == False
         isValid "\0" == False
Posix:   isValid "/random_ path:*" == True
Posix:   isValid x == not (null x)
```

The second and the fourth statements are contradictory...

If the fourth statement is true, we only have to look at what Path considers valid.

**Comment by sjakobi** [comment]
> > To speed up the generation it might also be possible to use System.FilePath.makeValid.

> Perhaps, but we have to make sure that all valid paths can still be generated.

From the  `makeValid` docs:

> does not change already valid FilePaths

As long as all parsing functions use `System.FilePath.isValid` we should be fine. That will change with #69 of course, but we can take care of this at that point.

> The second and the fourth statements are contradictory...

Maybe you're supposed to read them like a pattern match?!

**Comment by sjakobi** [comment]
With #79 merged validity-test runs in ~1.2 seconds on my machine now. That's good enough for me.

On Windows there's still a lot of room for improvement though. In particular generation of valid root directories could certainly be much faster.

**Comment by NorfairKing** [comment]
Done in recent changes.
